### PR TITLE
[CHERI-MIPS] Return structs that are pairs of capabilities in C3/C4

### PR DESCRIPTION
**Pull request by arichardson**
Since we are about to break the ABI anyway for two other changes I though I might as well fix this long-standing issue.

### CODE REVIEW

File: clang/lib/CodeGen/TargetInfo.cpp [hunk 1]
```diff
@@ -7656,6 +7656,52 @@ MipsABIInfo::returnAggregateInRegs(QualType RetTy, uint64_t Size) const {
...
+  } else if (const RecordType *RT = Ty->getAs<RecordType>()) {
+    if (!mipsCanReturnDirect(Ctx, RT->getDecl(), NumCaps, NumInts))
+      return false;
+  } else if (Ty->isConstantArrayType()) {
+    auto CAT = cast<ConstantArrayType>(Ty->getAsArrayTypeUnsafe());
+    if (CAT->getSize().ugt(2))
+      return false;
+    unsigned ArrayCaps = 0;
+    unsigned ArrayInts = 0;
+    if (!mipsCanReturnDirect(Ctx, CAT->getElementType(), ArrayCaps, ArrayInts))
```
Comment (jrtc27) [inline clang/lib/CodeGen/TargetInfo.cpp:7677]: This branch leaves `NumCaps` and `NumInt` as their initial values, so I think:
```
struct Foo {
    void * __capability x[2];
    int y;
}
```
would erroneously be returned in registers (or any struct that has any number of <=2-element int/cap arrays, and up to 2 other int/cap fields).

Comment (arichardson) [reply clang/lib/CodeGen/TargetInfo.cpp:7677]: Indeed! I meant to add the array values but somehow didn't include that...

Comment (jrtc27) [reply clang/lib/CodeGen/TargetInfo.cpp:7677]: Is there a reason you don't just pass NumCaps/NumInt directly as you do for records?..

Comment (arichardson) [reply clang/lib/CodeGen/TargetInfo.cpp:7677]: I meant to multiply it with the number of elements since it could be one or two.

**Comment by arichardson** [comment]
This has been merged to dev.

**Pull request closed without merging**